express/manifest.py

```diff
@@ -1,112 +1,113 @@
-"""
-Script for generating the dataset manifest that will be passed and updated through different
-components of the pipeline
-"""
+"""This module defines classes to represent an Express manifest."""
+
 import json
-from dataclasses import dataclass, field
-from enum import Enum
-from typing import List, Dict
+import pkgutil
+import typing as t
+from dataclasses import dataclass
 
-from dataclasses_json import dataclass_json
+import jsonschema.exceptions
+from jsonschema import Draft4Validator
 
+from express.exceptions import InvalidManifest
 
-class DataType(str, Enum):
-    """
-    Supported types for stored data.
-    """
 
-    PARQUET = "parquet"
-    BLOB = "blob"
+@dataclass
+class Field:
+    """Class representing a single field or column in an Express subset."""
 
-    @classmethod
-    def is_valid(cls, data_type: str) -> bool:
-        """Check if data type is valid"""
-        return data_type in cls.__members__.values()
+    name: str
+    type: str
 
 
-@dataclass
-class DataSource:
+class Subset:
     """
-    Information about the location and contents of a single data source.
+    Class representing an Express subset.
+
     Args:
-        location (str): the URI of the data root
-        type (Union[DataType, str]): the type of data, which determines how it's interpreted
-        extensions (List[str]): under what file extensions the data is stored.
-        n_files (int): how many files make up the data
-        n_items (int): how many items (e.g. distinct images, captions, etc.) are covered by the
-         data.
+        specification: The part of the manifest json representing the subset
+        base_path: The base path which the subset location is defined relative to
     """
 
-    location: str
-    type: DataType
-    extensions: List[str]
-    n_files: int = field(default_factory=int)
-    n_items: int = field(default_factory=int)
+    def __init__(self, specification: dict, *, base_path: str) -> None:
+        self._specification = specification
```

Comment:
```suggestion
        self.specification = specification
```
(just a question) can we just use regular variable names without _? I know they mean private but they hurt readability imo, just easier to have regular variable names. Also done like that in HF libraries

Reply:
Private variables and methods convey meaning on what should and should not be used by external classes. It also lets you limit the public interface of your class which you need to keep backward compatible.

I think the benefit of private variables is more than worth the underscore :slightly_smiling_face: